ml-algorithms/src/main/java/org/opensearch/ml/engine/memory/ConversationIndexMemory.java

```diff
@@ -0,0 +1,255 @@
+                    conversationId,
+                    parentId,
+                    traceNum,
+                    r.getId()
+                );
+        }, e -> { log.error("Failed to save interaction", e); }));
+    }
+
+    public void save(Message message, String parentId, Integer traceNum, String action, ActionListener listener) {
+        ConversationIndexMessage msg = (ConversationIndexMessage) message;
+        memoryManager
+            .createInteraction(
+                conversationId,
+                msg.getQuestion(),
+                null,
+                msg.getResponse(),
+                action,
+                Collections.singletonMap("None", "None"),
+                parentId,
+                traceNum,
+                listener
+            );
+    }
+
+    @Override
+    public void getMessages(String id, ActionListener listener) {
+        SearchRequest searchRequest = new SearchRequest();
+        searchRequest.indices(memoryMessageIndexName);
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
+        sourceBuilder.size(10000);
+        QueryBuilder sessionIdQueryBuilder = new TermQueryBuilder(SESSION_ID, id);
+
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        boolQueryBuilder.must(sessionIdQueryBuilder);
+
+        if (retrieveFinalAnswer) {
+            QueryBuilder finalAnswerQueryBuilder = new TermQueryBuilder("final_answer", true);
+            boolQueryBuilder.must(finalAnswerQueryBuilder);
+        }
+
+        sourceBuilder.query(boolQueryBuilder);
+        sourceBuilder.sort("created_time", SortOrder.ASC);
+        searchRequest.source(sourceBuilder);
+        client.search(searchRequest, listener);
+    }
+
+    public void getMessages(ActionListener listener) {
+        memoryManager.getFinalInteractions(conversationId, 10, listener);
+    }
+
+    @Override
+    public void clear() {}
+
+    @Override
+    public void remove(String id) {}
+
+    public static class Factory implements Memory.Factory<ConversationIndexMemory> {
+        private Client client;
+        private MLIndicesHandler mlIndicesHandler;
+        private String memoryMetaIndexName = ML_MEMORY_META_INDEX;
+        private String memoryMessageIndexName = ML_MEMORY_MESSAGE_INDEX;
+        private MLMemoryManager memoryManager;
+
+        public void init(Client client, MLIndicesHandler mlIndicesHandler, MLMemoryManager memoryManager) {
+            this.client = client;
+            this.mlIndicesHandler = mlIndicesHandler;
+            this.memoryManager = memoryManager;
+        }
+
+        @Override
+        public void create(Map<String, Object> map, ActionListener<ConversationIndexMemory> listener) {
+            if (map == null || map.isEmpty()) {
+                listener.onFailure(new IllegalArgumentException("Invalid input parameter for creating ConversationIndexMemory"));
+                return;
+            }
+
+            if (map.containsKey("memory_index_name")) {
```

Comment:
Create constant for "memory_index_name"?

Same comment for "memory_message_index_name" in line 180.

Reply:
done